### PR TITLE
Add a MLflow Tensorflow callback

### PR DESCRIPTION
**Pull request by chenmoneygithub**
## Related Issues/PRs

 #xxx

## What changes are proposed in this pull request?

Add callback `MLflowMetricsLoggingCallback` so that Tensorflow users can use MLflow as a callback. 

## How is this patch tested?

- [ ] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

## Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Users now can use MLflow as a Tensorflow callback in their `model.fit()` call. Please check [this colab](https://colab.research.google.com/drive/166MNmFKLbvVytUnL9zWkfnhsphm8kO-q?usp=sharing) for a demo.

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/gateway`: AI Gateway service, Gateway client APIs, third-party Gateway integrations
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/tracking/fluent.py [hunk 1]
```diff
@@ -1978,8 +1967,8 @@ def get_autologging_params(autolog_fn):
 
     def setup_autologging(module):
         try:
-            autolog_fn = LIBRARY_TO_AUTOLOG_FN[module.__name__]
-
+            autologging_params = None
+            autolog_fn = eval(LIBRARY_TO_AUTOLOG_FN[module.__name__])
```
Comment (harupy) [inline mlflow/tracking/fluent.py:1971]: Can we avoid using `eval` because it can lead to security vulnerabilities?

File: mlflow/tensorflow/callback.py
Comment (harupy) [inline mlflow/tensorflow/callback.py:1]: should this be documented? If yes, can you add a `.rst` file?

File: mlflow/tensorflow/callback.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    Args:
+        run: an 'mlflow.entities.run.Run' instance, the MLflow run.
+        log_every_epoch: bool, If True, log metrics every epoch. If False, log metrics every n
+            steps.
+        log_every_n_steps: int, log metrics every n steps. If None, log metrics every epoch.
+
+    Examples:
+        ```python
+        from tensorflow import keras
+        import mlflow
```
Comment (harupy) [inline mlflow/tensorflow/callback.py:22]: Does this code block render fine on the documentation?

File: mlflow/tensorflow/callback.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+            raise ValueError(
+                "`log_every_n_steps` must be specified if `log_every_epoch=False`, received"
+                "`log_every_n_steps=False` and `log_every_n_steps=None`."
+            )
+
+    def on_train_begin(self, logs=None):
+        """Log model architecture and optimizer configuration when training begins."""
+        config = self.model.optimizer.get_config()
+        for attribute in config:
+            log_param("optimizer_" + attribute, config[attribute])
```
Comment (harupy) [inline mlflow/tensorflow/callback.py:69]: Can we use `metrics_logger` for logging params?

File: pyproject.toml [hunk 4]
```diff
@@ -74,6 +74,8 @@ extend-exclude = [
 "dev/*" = ["T201", "PT018"]
 "examples/*" = ["T20"]
 "mlflow/*" = ["PT018"]
+"**/__init__.py" = ["F401"]
```
Comment (harupy) [inline pyproject.toml:77]: Do we need this? We can avoid `F401` by adding names `__all__` in `__init__.py`.

File: mlflow/tensorflow/callback.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+                data,
+                label,
+                batch_size=4,
+                epochs=2,
+                callbacks=[MLflowMetricsLoggingCallback(run)],
+            )
+        ```
+    """
+
+    def __init__(self, run, log_every_epoch=True, log_every_n_steps=None):
```
Comment (harupy) [inline mlflow/tensorflow/callback.py:54]:
```suggestion
    def __init__(self, run_id, log_every_epoch=True, log_every_n_steps=None):
```

Can we use `run_id` instead of `run`?

File: mlflow/tensorflow/callback.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+        if not self.log_every_epoch:
+            return
+        self.metrics_logger.record_metrics(logs, epoch)
+
+    def on_batch_end(self, batch, logs=None):
+        """Log metrics at the end of each batch with user specified frequency."""
+        if self.log_every_n_steps is None:
+            return
+        if (batch + 1) % self.log_every_n_steps == 0:
+            self.metrics_logger.record_metrics(logs, batch)
```
Comment (WeichenXu123) [inline mlflow/tensorflow/callback.py:87]: Q:
In `on_epoch_end`, the step is epoch index, but in `on_batch_end`, the step is the batch index, but they have the overlapped metric keys (I think they have, could you list typical `logs` argument values for a model?), then it causes confusion

Comment (WeichenXu123) [reply mlflow/tensorflow/callback.py:87]: Q:
Shall we support method like `on_test_batch_end` ?

File: mlflow/tracking/fluent.py (hunk 1 above)
Comment (WeichenXu123) [reply mlflow/tracking/fluent.py:1971]: Master code `autolog_fn = LIBRARY_TO_AUTOLOG_FN[module.__name__]` should work, what's the issue you want to address by adding `eval` ?

File: mlflow/tensorflow/callback.py (hunk 3 above)
Comment (WeichenXu123) [reply mlflow/tensorflow/callback.py:69]: `metrics_logger` seemingly is designed for logging metrics but not param ?

File: mlflow/tracking/fluent.py (hunk 1 above)
Comment (chenmoneygithub) [reply mlflow/tracking/fluent.py:1971]: the problem is `autolog()` imports all flavor modules, which is violating the lazy loading logic. 

`eval` is unsafe for inputs, but here it only takes in hardcoded string of module path.

File: mlflow/tensorflow/callback.py (hunk 3 above)
Comment (chenmoneygithub) [reply mlflow/tensorflow/callback.py:69]: yea I am not sure about this actually, basically I am reusing the logic from autologging modules.

File: mlflow/tensorflow/callback.py
Comment (chenmoneygithub) [reply mlflow/tensorflow/callback.py:1]: sorry what's the `.rst` file? can I have an example? thanks!

File: mlflow/tensorflow/callback.py (hunk 2 above)
Comment (chenmoneygithub) [reply mlflow/tensorflow/callback.py:22]: there should be a way, but I am done to be consistent with other modules. changed.

File: pyproject.toml (hunk 4 above)
Comment (chenmoneygithub) [reply pyproject.toml:77]: using `__all__` sg!

File: mlflow/tensorflow/callback.py (hunk 5 above)
Comment (chenmoneygithub) [reply mlflow/tensorflow/callback.py:54]: honestly I have been back and forth about `run_id` vs `run` in the constructor. I think `run` is slightly better because users can just do `callback = MLflowCallback(run)` instead of `callback = MLflowCallback(run.info.run_id)`, we are hiding the complexity under the hood.

File: mlflow/tensorflow/callback.py (hunk 6 above)
Comment (chenmoneygithub) [reply mlflow/tensorflow/callback.py:87]: > In on_epoch_end, the step is epoch index, but in on_batch_end, the step is the batch index
Good call, and yes, it will be confusing to log both. I am letting users choose either log per epoch or log per N steps, not both. 

> Shall we support method like on_test_batch_end ?
Good idea, will add code for it.

File: mlflow/tensorflow/callback.py (hunk 3 above)
Comment (harupy) [reply mlflow/tensorflow/callback.py:75]: Ah you're right. It only supports logging metrics.

File: mlflow/tracking/fluent.py (hunk 1 above)
Comment (harupy) [reply mlflow/tracking/fluent.py]: The issue with `eval` is that that makes us wonder "Is this really safe?". Can we use `importlib` instead?

File: pyproject.toml [hunk 7]
```diff
@@ -74,7 +74,7 @@ extend-exclude = [
 [tool.ruff.per-file-ignores]
 "dev/*" = ["T201", "PT018"]
 "examples/*" = ["T20"]
-"mlflow/*" = ["PT018"]
+"mlflow/*" = ["PT018"]
```
Comment (harupy) [inline pyproject.toml:77]:
```suggestion
"mlflow/*" = ["PT018"]
```

File: mlflow/tensorflow/callback.py [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+
+    def on_test_end(self, logs=None):
+        """Log validation metrics at validation end."""
+        if logs is None:
+            return
+        metrics = {}
+        for k, v in logs.items():
+            metrics["validation_" + k] = v
```
Comment (harupy) [inline mlflow/tensorflow/callback.py:101]: /ai rewrite this code using dict comprehension

File: mlflow/tensorflow/callback.py (hunk 5 above)
Comment (harupy) [reply mlflow/tensorflow/callback.py:55]: Makes sense, let's use `run`.

File: mlflow/tensorflow/callback.py [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+
+
+class MLflowLoggingCallback(keras.callbacks.Callback):
+    """Callback for logging Tensorflow training metrics to MLflow.
+
+    This callback logs training metrics every epoch or every n steps (defined by the user) to
+    MLflow.
+
+    Args:
+        run: an 'mlflow.entities.run.Run' instance, the MLflow run.
```
Comment (harupy) [inline mlflow/tensorflow/callback.py:14]:
```suggestion
        run: an `mlflow.entities.run.Run` instance, the MLflow run.
```

File: mlflow/tensorflow/callback.py [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+            raise ValueError(
+                "`log_every_n_steps` must be specified if `log_every_epoch=False`, received"
+                "`log_every_n_steps=False` and `log_every_n_steps=None`."
+            )
+
+    def on_train_begin(self, logs=None):
+        """Log model architecture and optimizer configuration when training begins."""
+        config = self.model.optimizer.get_config()
+        for attribute in config:
+            log_param("optimizer_" + attribute, config[attribute])
```
Comment (harupy) [inline mlflow/tensorflow/callback.py:75]:
```suggestion
        log_params({f"optimizer_{k}": v for k, v in config.items()})
```

File: mlflow/tensorflow/callback.py [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+
+class MLflowLoggingCallback(keras.callbacks.Callback):
+    """Callback for logging Tensorflow training metrics to MLflow.
+
+    This callback logs training metrics every epoch or every n steps (defined by the user) to
+    MLflow.
+
+    Args:
+        run: an 'mlflow.entities.run.Run' instance, the MLflow run.
+        log_every_epoch: bool, If True, log metrics every epoch. If False, log metrics every n
```
Comment (harupy) [inline mlflow/tensorflow/callback.py:15]: Can we use `log_every_n_epochs` instead?

File: mlflow/tensorflow/callback.py [hunk 12]
```diff
@@ -0,0 +1,102 @@
+from tensorflow import keras
+
+from mlflow import log_param, log_text
+from mlflow.utils.autologging_utils import BatchMetricsLogger
+
+
+class MLflowLoggingCallback(keras.callbacks.Callback):
```
Comment (harupy) [inline mlflow/tensorflow/callback.py:7]:
```suggestion
class MLflowTensorflowCallback(keras.callbacks.Callback):
```

?

Comment (harupy) [reply mlflow/tensorflow/callback.py:7]: or

```suggestion
class MLflowCallback(keras.callbacks.Callback):
```

? wandb uses this convention: https://docs.wandb.ai/guides/integrations/xgboost#get-started

File: mlflow/tensorflow/callback.py [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+    """Callback for logging Tensorflow training metrics to MLflow.
```
Comment (harupy) [inline mlflow/tensorflow/callback.py:8]: Can we be more precise? We log params, metrics, and model summary.

**Review by harupy — APPROVED** [review]
Left a few more comments, otherwise LGTM!

File: mlflow/tensorflow/callback.py (hunk 12 above)
Comment (harupy) [reply mlflow/tensorflow/callback.py:7]: or

```suggestion
class TensorflowCallback(keras.callbacks.Callback):
```

File: mlflow/tensorflow/callback.py (hunk 11 above)
Comment (chenmoneygithub) [reply mlflow/tensorflow/callback.py:15]: It's very rare that users have a callback working per N epochs, I think it makes more sense to just do every epoch.
